content/things/rust-to-webassembly.md

```diff
@@ -0,0 +1,637 @@
+instance.exports.add(40, 2) // returns 42
+```
+
+And suddenly, we have pretty much all the power of Rust at our fingertips to write WebAssembly. 
+
+> **Streaming:** On the web it is advisable to _not_ call `r.arrayBuffer()` and instead pass the `Response` straight to `WebAssembly.instantiateStreaming`. This allows the WebAssembly compiler to start compiling while the file is being downloaded.
+
+Special care needs to be taken with functions at the module boundary (i.e. the ones you call from JavaScript). It’s best to stick to types that map cleanly to [WebAssembly types] (like `i32` or `f64`). If you use higher-level types like arrays, slices, or even owned types like `String`, it will compile, but yield a unexpected function signature and will generally become harder to use. More on that later!
+
+### Importing
+
+One important part of WebAssembly is its sandbox. It ensures that the code running in the WebAssembly VM gets no access to anything in the host environment apart from the functions that were explicitly passed into the sandbox via the imports object.
+
+Let’s say we want to generate some random numbers in our Rust code. We could pull in the `rand` Rust crate, but why ship code for something if we can grab the existing implementation for free. To make that work, we need to declare that our WebAssembly module expects an import.
+
+|||codediff|rust
++ #[link(wasm_import_module = "Math")]
++ extern "C" {
++     #[link_name = "random"]
++     fn random() -> f64;
++ }
+  
+  #[export_name = "add"]
+  pub fn add(left: f64, right: f64) -> f64 {
+-     left + right  
++     left + right + unsafe { random() }
+  }
+|||
+
+`extern "C"` blocks declare functions that we assume to be provided by ”someone else” during linking. This is usually how you link against C libraries in Rust, but the mechanism works for WebAssembly as well. However, external functions are always implicitly unsafe, as the compiler can’t make any safety guarantees for non-Rust functions that are not present. As a result, we need to wrap in `unsafe { ... }` blocks whenever we call them.
+
+This code will compile, but our JavaScript code that instantiates the module now throws an error. If a WebAssembly module expects imports, they _must_ be present on the imports object. 
+
+The imports _object_ is a dictionary of import _modules_, that each are dicitonary of import _items_. In our Rust code we declared ab import module with the name `"Math"`, and expect a function called `"random"` to be present in that module. These values have of course been carefully chosen so that we can just pass in the entire `Math` object.
+
+|||codediff|js
+  const importObj = {
++   Math
+  };
+|||
+
+To avoid having to sprinkle `unsafe { ... }` everywhere, it is often desirable to write wrapper functions that restore the safety invariants of Rust.
+
+```rust
+#[link(wasm_import_module = "Math")]
+extern "C" {
+    #[link_name = "random"]
+    fn random_unsafe() -> f64;
```

Comment:
Rather than renaming functions, I'd suggest wrapping the `extern` in a `mod`, e.g. `mod Math_unsafe` or `mod Math_js` or just `mod Math`. That way, you don't need to introduce `link_name` for renamings, and, as a pattern, it scales better if user wants to add more functions, since they can just do `Math_unsafe::random()`, `Math_unsafe::pow()` etc. in their safe wrappers.

Reply:
Love it.